thresholdout4real_data/refactor_general_simulation.R

```diff
@@ -0,0 +1,266 @@
+      # model selection (select the best subset of features)
+      th_auc_vec[model_ind] <- temp$thresholdout_auc
+      thresholdout_params <- temp$params
+    }
+    auc <- mutate(auc, thresholdout_auc = th_auc_vec)
+
+    # package results
+    # choose the best up till now
+    best_model_ind <- which.max(auc$thresholdout_auc)  # use thresholdout result as model selection indicator
+    # (the recorded thresholdout_auc is not useful for model evaluation,
+    # because it captures an instantiation where the added noise happened
+    # to be extremely large and positive. So recalculate the Thresholdout
+    # score for the best model)
+    temp <- thresholdout_auc(thresholdout_params = thresholdout_params,
+                             train_auc = auc$repeatedcv_auc[best_model_ind],
+                             holdout_auc = auc$holdout_auc[best_model_ind])
+    auc$thresholdout_auc[best_model_ind] <- temp$thresholdout_auc
+    thresholdout_params <- temp$params
+    features_to_keep <- union(features_to_keep, selected_features[[best_model_ind]])
+    num_features_by_round <- c(num_features_by_round, length(features_to_keep))
+    holdout_access_by_round <- c(holdout_access_by_round, nrow(auc) + 1)
+    cum_budget_decrease_by_round <- c(cum_budget_decrease_by_round,
+                                      thresholdout_params$budget_utilized)
+    # combine results from previous rounds
+    auc[, "train_auc"]
+    auc[, "test_auc"]
+    auc[, "holdout_auc"]  # auc has the number of rows which is the subset feature combination
+    print(best_model_ind)
+    print(auc_by_round_df)
+    auc_by_round_df <- bind_rows(auc_by_round_df,
+                                 mutate(auc[best_model_ind, ], round = round_ind))
+    # auc_by_round_df$dataset
+  }
+
+  auc_by_round_df <- auc_by_round_df %>% gather(dataset, auc, -round, -n_train)
+
+  return(list(selected_features = features_to_keep,
+              num_features_by_round = num_features_by_round,
+              auc_by_round_df = auc_by_round_df,
+              holdout_access_by_round = holdout_access_by_round,
+              cum_holdout_access = cumsum(holdout_access_by_round),
+              cum_budget_decrease_by_round = cum_budget_decrease_by_round))
+}
+
+
+
+
+#' @title 
+#' @description
+#' @param task a mlr::task
+#' @return list(n_train = n_train, x_train_total = x_train_total, y_train_total = y_train_total, x_holdout = x_holdout, y_holdout = y_holdout, x_test = x_test, y_test = y_test, tname = tname, bname = bname, p = p)
+#' @examples 
+#' demo_data_fun(mlr::sonar.task)
+demo_data_fun = function(instance = NULL, conf = NULL) {
+  if (is.null(instance)) {
+    task = mlr::sonar.task
+  } else {
+    stop("demo_data_fun could only accept instance = NULL")
+  }
+  tname = mlr::getTaskTargetNames(task)
+  bname = mlr::getTaskDesc(task)$negative
+  p = mlr::getTaskNFeats(task)
+  n_all = mlr::getTaskSize(task)
+
+  n_train_total = 0.5 * n_all
```

Comment:
I haven't tried it, but intuitively it seems to me that `n_train_total` is set too large. I think that by setting `n_train_total` to be half of all available data, the variability between the different simulation runs will be much reduced. That is because training datasets between different simulation runs will overlap to a large extent, and therefore different runs of `run_sim` will behave very similarly. That is, we would gain less information about the behaviour of ThresholdoutAUC by running the simulation multiple times, compared to the situation where the datasets are not overlapping as much between simulation runs.
Of course, how `n_train_total` is set should be determined by what you want to achieve by running this simulation.